test/extensions/dynamic_modules/test_data/rust/http_test.rs

```diff
@@ -0,0 +1,79 @@
+use super::*;
+use envoy_proxy_dynamic_modules_rust_sdk::{EnvoyBuffer, EnvoyHttpFilter};
+
+#[test]
+fn test_header_callbacks_filter_on_request_headers() {
+  let mut f = HeaderCallbacksFilter {};
+  let mut envoy_filter = EnvoyHttpFilter::default();
+
+  envoy_filter
+    .expect_get_request_header_value()
+    .withf(|name| name == "single")
+    .returning(|_| Some(EnvoyBuffer::new("value".as_ptr(), "value".len())))
+    .once();
+
+  envoy_filter
+    .expect_get_request_header_value()
+    .withf(|name| name == "non-exist")
+    .returning(|_| None)
+    .once();
+
+  envoy_filter
+    .expect_get_request_header_values()
+    .withf(|name| name == "multi")
+    .returning(|_| {
+      let value1 = EnvoyBuffer::new("value1".as_ptr(), "value1".len());
+      let value2 = EnvoyBuffer::new("value2".as_ptr(), "value2".len());
+      vec![value1, value2]
+    })
+    .once();
+
+  envoy_filter
+    .expect_get_request_header_values()
+    .withf(|name| name == "non-exist")
+    .returning(|_| vec![])
+    .once();
+
+  envoy_filter
+    .expect_set_request_header()
+    .withf(|name, value| name == "new" && value == b"value")
+    .return_const(true)
+    .once();
+
+  envoy_filter
+    .expect_get_request_header_value()
+    .withf(|name| name == "new")
+    .returning(|_| {
+      const VALUE: &str = "value";
+      Some(EnvoyBuffer::new(VALUE.as_ptr(), VALUE.len()))
+    })
+    .once();
+
+  envoy_filter
+    .expect_get_request_headers()
+    .returning(|| {
+      let single = (
+        EnvoyBuffer::new("single".as_ptr(), "single".len()),
```

Comment:
It would be much better to make EnvoyBuffer a trait and then have an implementation for Envoy and then possibly an owned version used for testing. This is fairly risky the way you have this since you can easily create undefined behavior. Alternatively, hav a constructor version that accepts `'static` lifetime strings as arguments and then an unsafe constructor version that the real Envoy code uses.

Reply:
https://github.com/envoyproxy/envoy/pull/37753/commits/c2098852b831c212ce29c6793723bca1e6aabb60 added `'static` constructor - I will revisit this when/after someone complains about this only being 'static later